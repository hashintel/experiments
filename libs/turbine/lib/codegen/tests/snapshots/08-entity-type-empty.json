[
  {
    "$id": "http://localhost:3000/@alice/types/entity-type/country/v/1",
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
    "allOf": [],
    "description": null,
    "kind": "entityType",
    "links": {},
    "properties": {},
    "required": [],
    "title": "Country",
    "type": "object"
  }
]
