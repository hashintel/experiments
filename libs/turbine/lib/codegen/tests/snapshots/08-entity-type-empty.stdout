localhost_3000/alice/entity/country.rs

use alloc::string::String;
use core::marker::PhantomData;

use error_stack::{Report, Result, ResultExt as _};
use hashbrown::HashMap;
use serde::Serialize;
use turbine::{
    entity::Entity, url, BaseUrl, EntityLink, EntityProperties, EntityType, EntityTypeMut,
    EntityTypeRef, GenericEntityError, OptionalEntityLink, PropertyType as _, PropertyTypeMut as _,
    PropertyTypeRef as _, Type, TypeMut, TypeRef, TypeUrl, VersionedUrlRef,
};
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Properties;
turbine::serialize_compat!(Properties);
impl Properties {
    fn try_from_value(
        mut properties: HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        Ok(Self)
    }

    fn as_mut(&mut self) -> PropertiesMut<'_> {
        PropertiesMut(PhantomData)
    }

    fn as_ref(&self) -> PropertiesRef<'_> {
        PropertiesRef(PhantomData)
    }
}
///Country
#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Country {
    pub properties: Properties,
}
pub type CountryV1 = Country;
impl TypeUrl for Country {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl Type for Country {
    type Mut < 'a > = CountryMut < 'a > where Self : 'a ;
    type Ref < 'a > = CountryRef < 'a > where Self : 'a ;

    fn as_mut(&mut self) -> Self::Mut<'_> {
        CountryMut {
            properties: self.properties.as_mut(),
        }
    }

    fn as_ref(&self) -> Self::Ref<'_> {
        CountryRef {
            properties: self.properties.as_ref(),
        }
    }
}
impl EntityType for Country {
    type Error = GenericEntityError;

    fn try_from_entity(value: Entity) -> Option<Result<Self, Self::Error>> {
        if Self::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = Properties::try_from_value(value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl EntityProperties for Country {
    type Properties = Properties;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for Country {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PropertiesRef<'a>(pub PhantomData<&'a ()>);
turbine::serialize_compat!(PropertiesRef<'a>);
impl<'a> PropertiesRef<'a> {
    fn try_from_value(
        properties: &'a HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        Ok(Self(PhantomData))
    }

    fn into_owned(self) -> Properties {
        Properties
    }
}
///Country
#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CountryRef<'a> {
    pub properties: PropertiesRef<'a>,
}
pub type CountryV1Ref<'a> = CountryRef<'a>;
impl TypeUrl for CountryRef<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl TypeRef for CountryRef<'_> {
    type Owned = Country;

    fn into_owned(self) -> Self::Owned {
        Country {
            properties: self.properties.into_owned(),
        }
    }
}
impl<'a> EntityTypeRef<'a> for CountryRef<'a> {
    type Error = GenericEntityError;

    fn try_from_entity(value: &'a Entity) -> Option<Result<Self, Self::Error>> {
        if Self::Owned::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = PropertiesRef::try_from_value(&value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl<'a> EntityProperties for CountryRef<'a> {
    type Properties = PropertiesRef<'a>;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for CountryRef<'_> {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}
#[derive(Debug)]
pub struct PropertiesMut<'a>(pub PhantomData<&'a ()>);
turbine::serialize_compat!(PropertiesMut<'a>);
impl<'a> PropertiesMut<'a> {
    fn try_from_value(
        properties: &'a mut HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        Ok(Self(PhantomData))
    }

    fn into_owned(self) -> Properties {
        Properties
    }
}
///Country
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CountryMut<'a> {
    pub properties: PropertiesMut<'a>,
}
pub type CountryV1Mut<'a> = CountryMut<'a>;
impl TypeUrl for CountryMut<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl TypeMut for CountryMut<'_> {
    type Owned = Country;

    fn into_owned(self) -> Self::Owned {
        Country {
            properties: self.properties.into_owned(),
        }
    }
}
impl<'a> EntityTypeMut<'a> for CountryMut<'a> {
    type Error = GenericEntityError;

    fn try_from_entity(value: &'a mut Entity) -> Option<Result<Self, Self::Error>> {
        if Self::Owned::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = PropertiesMut::try_from_value(&mut value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl<'a> EntityProperties for CountryMut<'a> {
    type Properties = PropertiesMut<'a>;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for CountryMut<'_> {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}
